pkg/controller/replica.go

```diff
@@ -300,13 +313,34 @@ func NewReplicaController(mgr manager.Manager, controllerSharedState *SharedStat
 			return false
 		},
 		UpdateFunc: func(e event.UpdateEvent) bool {
-			return true
+			objOld, okOld := e.ObjectOld.(*azdiskv1beta2.AzVolumeAttachment)
+			objNew, okNew := e.ObjectNew.(*azdiskv1beta2.AzVolumeAttachment)
+			if okOld && okNew {
+				demoteRequestUpdate := !isDemotionRequested(objOld) && isDemotionRequested(objNew)
+				promoteRequestUpdate := !isPromotionRequested(objOld) && isPromotionRequested(objNew)
+				attachRequestUpdate := !volumeAttachRequested(objOld) && volumeAttachRequested(objNew)
+				detachRequestUpdate := !volumeDetachRequested(objOld) && volumeDetachRequested(objNew)
+				attachFailedUpdate := objOld.Status.State != azdiskv1beta2.AttachmentFailed && objNew.Status.State == azdiskv1beta2.AttachmentFailed
+				detachFailedUpdate := objOld.Status.State != azdiskv1beta2.DetachmentFailed && objNew.Status.State == azdiskv1beta2.DetachmentFailed
+
+				deleteRequestedOld, _ := objectDeletionRequested(objOld)
+				deleteRequestedNew, _ := objectDeletionRequested(objNew)
+				deleteTimestampUpdate := !deleteRequestedOld && deleteRequestedNew
+
+				updateEventsForPrimary := (objNew.Spec.RequestedRole == azdiskv1beta2.PrimaryRole) && (attachRequestUpdate || promoteRequestUpdate || detachRequestUpdate)
+				updateEventsForReplica := (objNew.Spec.RequestedRole == azdiskv1beta2.ReplicaRole) && (demoteRequestUpdate || attachFailedUpdate || deleteTimestampUpdate || detachFailedUpdate)
+
+				if updateEventsForPrimary || updateEventsForReplica {
+					return true
+				}
```

Comment:
This can be summed to `return updateEventsForPrimary || updateEventsForReplica` :)

Reply:
Nice suggestion!